fpdf/line_break.py

```diff
@@ -9,47 +17,144 @@
+        self.k = k
 
     def __repr__(self):
-        return f"Fragment(style={self.style}, underline={self.underline}, characters={self.characters})"
+        gstate = self.graphics_state.copy()
+        if "current_font" in gstate:
+            del gstate["current_font"]  # TMI
+        return (
+            f"Fragment(characters={self.characters},"
+            f" graphics_state={gstate}, k={self.k})"
+        )
+
+    @property
+    def font(self):
+        return self.graphics_state["current_font"]
 
-    @classmethod
-    def from_string(cls, string: str, style: str, underlined: bool):
-        return cls(style, underlined, list(string))
+    @font.setter
+    def font(self, v):
+        self.graphics_state["current_font"] = v
 
-    def trim(self, index: int):
-        self.characters = self.characters[:index]
+    @property
+    def unicode_font(self):
+        return self.font.get("type") == "TTF"
+
+    @property
+    def font_style(self):
+        return self.graphics_state["font_style"]
+
+    @property
+    def font_family(self):
+        return self.graphics_state["font_family"]
+
+    @property
+    def font_size_pt(self):
+        return self.graphics_state["font_size_pt"]
+
+    @property
+    def font_size(self):
+        return self.graphics_state["font_size_pt"] / self.k
+
+    @property
+    def font_stretching(self):
+        return self.graphics_state["font_stretching"]
+
+    @property
+    def char_spacing(self):
+        return self.graphics_state["char_spacing"]
+
+    @property
+    def underline(self):
+        return self.graphics_state["underline"]
 
     @property
     def string(self):
         return "".join(self.characters)
 
+    def trim(self, index: int):
+        self.characters = self.characters[:index]
+
     def __eq__(self, other: Any):
         return (
             self.characters == other.characters
-            and self.style == other.style
-            and self.underline == other.underline
+            and self.graphics_state == other.graphics_state
+            and self.k == other.k
         )
 
+    def get_width(self, start: int = 0, end: int = None, chars: str = None):
+        """
+        Return the witdth of the string with the given font/size/style/etc.
+
+        Args:
+            start (int): Index of the start character. Default start of fragment.
+            end (int): Index of the end character. Default end of fragment.
+            chars (str): Specific text to get the width for (not necessarily the
+                same as the contents of the fragment). If given, this takes
+                precedence over the start/end arguments.
+        """
+
+        def char_width(char):
+            try:
+                width = self.font["cw"][char]
+            except (IndexError, KeyError):
+                width = self.font.get("desc", {}).get("MissingWidth") or 500
+            if width == 65535:
+                return 0
+            return width
```

Comment:
In my pull request, I changed `self.font["cw"]` (previously just a `list`) to a `defaultdict` with a `default_width`, so we could remove the `try... except` here, what do you think?

I also moved 
```python
if width == 65535:
   width = 0
```
in `.add_font()` when building the dictionary, in this way we could simplify the method in just a one-line: 
```python
return self.font["cw"][char]
```

Reply:
To be honest, I didn't really look at the reason for this `width == 65535` condition, but just used it the way I had found it in fpdf.py. You had a closer look at the details there, so I'll trust your judgement (and test cases).

Which ever of our PRs gets accepted last will have to transfer your change over here.

Reply:
> the reason for this width == 65535 condition

Unfortunately, I don't know it either. I tried to remove it, but some tests broke, so I moved it when `font["cw"]` is initialized in `add_font()`